### PR TITLE
Fix: Duplicate fragments in populated documents

### PR DESCRIPTION
**Pull request by imranolas**
Where the mutation body uses and declares a fragment named the same as one found in an active selection it was incorrectly added as duplicate. This change checks the document for existing named fragments before adding them.

Given a query
```gql
query {
  todos {
    ...TodoFragment
    creator {
      ...CreatorFragment
    }
  }
}

fragment TodoFragment on Todo {
  id
  text
}

fragment CreatorFragment on User {
  id
  name
}
```

and mutation using both `@populate` and a named fragment
```gql
mutation MyMutation {
  addTodo @populate
    ...TodoFragment
  }
}
      
fragment TodoFragment on Todo {
  id
  text
}
```

The `TodoFragment` should appear only once in the populated mutation.

## Additional changes
- Renamed some variables for clarity
- Moved the snapshots to inline for ease of readability. Reviewers may find it simpler to step through commits.

### CODE REVIEW

File: src/populateExchange.test.ts [hunk 1]
```diff
@@ -34,8 +34,8 @@ const schemaDef = `
   union UnionType = User | Todo
 
   type Query {
-    todos: [Todo]
-    users: [User]
+    todos: [Todo!]
```
Comment (imranolas) [inline src/populateExchange.test.ts]: Added some nullablity to the schema for broader test coverage. This should also test the cases fix by #150

File: src/populateExchange.test.ts [hunk 2]
```diff
@@ -145,13 +173,21 @@ describe('on (query w/ fragment) -> mutation', () => {
       query {
         todos {
           ...TodoFragment
+          creator {
+            ...CreatorFragment
```
Comment (imranolas) [inline src/populateExchange.test.ts]: Extended this test to trigger the case covered by this fix

File: src/populateExchange.ts [hunk 3]
```diff
@@ -251,6 +256,14 @@ export const addFragmentsToQuery = (
         },
       },
       Document: {
+        enter: node => {
+          node.definitions.reduce((set, definition) => {
```
Comment (imranolas) [inline src/populateExchange.ts]: Adds the existing fragment definitions to a Set for subsequent checks

File: src/populateExchange.ts [hunk 4]
```diff
@@ -167,7 +167,7 @@ export const extractSelectionsFromQuery = (
 export const addFragmentsToQuery = (
   schema: GraphQLSchema,
   query: DocumentNode,
-  typeFragments: TypeFragmentMap,
+  activeTypeFragments: TypeFragmentMap,
```
Comment (andyrichardson) [inline src/populateExchange.ts]: :heavy_plus_sign: for this - We should probably do this for [here also](https://github.com/FormidableLabs/urql-exchange-graphcache/blob/8f0cbc326964ce0c090c4ba6b965285841834ad3/src/populateExchange.ts#L41)

File: src/populateExchange.ts (hunk 3 above)
Comment (andyrichardson) [reply src/populateExchange.ts]: I'd go for a `forEach` as we're mutating the set and it's already in scope

Comment (kitten) [reply src/populateExchange.ts]: We’ll have to go for a `forEsch` here actually as it’s a `Set` and IE11 doesn’t implement the iterable spec :)

Comment (imranolas) [reply src/populateExchange.ts]: It's not iterating on a Set – it's reducing into one

Comment (kitten) [reply src/populateExchange.ts]: oh true! mahbad

**Pull request merged**